content/blog/2025-09-15-hash-fields-expiration/index.md

```diff
@@ -0,0 +1,225 @@
+In practice, the overhead ranged between 16 and 29 bytes per field.
+
+![illustration](hfe-benchmark-memory.png)
+
+The higher end of this range primarily affects small hashes, where compact encodings
+like listpack are avoided when volatile fields are present.
+
+### Command Performance
+
+Next, we benchmarked common hash commands both with and without field expirations.
+The results showed no measurable performance regression and throughput remained stable when expirations were added.
+
+![illustration](hfe-benchmark-hash-commands.png)
+
+We also benchmarked the new expiration-aware commands (e.g., `HSETEX`), confirming that their performance is on par with traditional hash operations.
+
+![illustration](hfe-benchmark-new-commands.png)
+
+### Active Expiration Efficiency
+
+The design goal of the vset was to enable efficient background deletion of expired fields.
+To test this, we preloaded 10 million fields with TTLs.
+We distributed these fields across varying numbers of hash objects to see how object size influences expiration.
+During the load phase, we disabled the expiration job using the [`DEBUG`](/commands/debug/), then re-enabled it once all fields had expired.
+
+The following chart shows the time it took the expiration cron job complete the full deletion of all the 10M fields.
+
+![illustration](hfe-benchmark-new-commands.png)
+
+The results revealed that expiration time depends not just on the number of fields, but also on how they are distributed across objects.
+Smaller hashes tend to fit into CPU caches, so random field deletions remain cache-friendly.
+Very large hashes, however, cannot fit entirely in cache, which means more expensive memory lookups during expiration.
+
+We also simulated a more realistic scenario: *expiring data during continuous ingestion.*
+Using the 'valkey-benchmark' executing the new [`HSETEX`](/commands/hsetex) command, we continuously
+inserted 10 million fields with 10-second TTLs, while the active expiration job ran in the background.
+
+This setup maintained a constant pool of fields at different stages of their lifecycle — some fresh, some nearing expiration, some ready to be reclaimed.
+We then tracked memory usage over a 5-minute period.
+
+![illustration](hfe-benchmark-active-expiry.png)
+
+The results aligned with our theoretical expectation:
+
+```
+Memory = (Injection Throughput) x (AVG TTL) x (AVG Item memory)
+```
+
+In our experiment:
+
+* Injection rate: 300K commands/sec
+
+* Average TTL: 10 seconds
+
+* Base item size: ~61 bytes
+
+* Additional expiration overhead: ~19 bytes
+
+This yields an expected memory footprint of ~230MB:
+
+```
+300K × 10 × 80B = 230MB
+```
+
+The observed memory matched this estimate closely, demonstrating that the active expiration mechanism
+is able to keep up with load and prevent memory from spiking unexpectedly.
+
+## Conclusion
+
+The benchmarks demonstrate that field-level expirations can be added to Valkey without compromising performance or stability.
```

Comment:
```suggestion
The benchmarks demonstrate that field-level expirations can be added to Valkey without compromising performance, memory efficiency, or latency.
```
I don't see mentions of stability anywhere.

Reply:
Well. We did not verify that in benchmarking :)  I wonder how can I claim it without address the tests?

Reply:
Oh. right. I agree.